### DISCUSSION TITLE
Unable to Receive/Sniff LoRaWAN Packets with CATSniffer V3.1 from Heltec LoRa V3

### DISCUSSION THREAD

**Discussion by joterr** [jgromes/RadioLib#915] [Q&A] [answered]
Hello,

I am trying to sniff LoRaWan packets with a CATSniffer V3.1  (RP2040 Pico and SX1262 via SPI) and the following code.

The Heltec LoRa V3 works with the LoRaWAN implementation by Heltec (LoRaWan_APP) to send LoRaWAN packets.
A Milesight Gateway can receive these packets successfully - but the CATSniffer with RadioLib is not getting the data.

What do I miss?

Using `int state = radio.receive(str);` I get `RADIOLIB_ERR_NONE`, but no data is displayed. Should I use the callback / receiveStart() option? 

```
Initializing with ...	Freq: 868.10	SF: 7
success!
Waiting for incoming transmission ... 
[[[NEW PACKET]]]
Time:			84198
Data:			
Length:		23 bytes
IRQ:			0
RSSI:			-34.00 dBm
SNR:			13.00 dB
FreqError:		5959.75 Hz
Freq:			868.10 Hz
SF:			7
```

Thank you,
Jannik

Ino file from Arduino IDE:

```
#define SERIALCOMMAND_HARDWAREONLY

#include <SerialCommand.h>
#include <RadioLib.h>

#define CTF1 8
#define CTF2 9
#define CTF3 10

#define freqSize 3
#define sfSize 1

SerialCommand SCmd;

float frequency[freqSize] = { 868.1, 868.3, 868.5 }; // , 868.3, 868.5
int sf[sfSize] = { 7 };

int freqCounter = 0;
int sfCounter = 0;

// SX1262 has the following connections:
// NSS pin:   17
// DIO1 pin:  5
// NRST pin:  24
// BUSY pin:  4
SX1262 radio = new Module(17, 5, 24, 4);


void setup() {
  Serial.begin(921600);
  while (!Serial)
    ;

  pinMode(CTF1, OUTPUT);
  pinMode(CTF2, OUTPUT);
  pinMode(CTF3, OUTPUT);

  digitalWrite(CTF1, HIGH);
  digitalWrite(CTF2, LOW);
  digitalWrite(CTF3, LOW);

  Serial.print(F("Initializing with ..."));
  Serial.print(F("\tFreq: "));
  Serial.print(frequency[freqCounter]);
  Serial.print(F("\tSF: "));
  Serial.println(sf[sfCounter]);

  int state = radio.begin(frequency[freqCounter], 125.0, sf[sfCounter], 7, RADIOLIB_SX126X_SYNC_WORD_PUBLIC, 10, 8, 0, false);

  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true)
      ;
  }
  Serial.println(F("Waiting for incoming transmission ... "));
}
void loop() {
  String str;
  int state = radio.receive(str);

  if (state == RADIOLIB_ERR_NONE) {
    // packet was successfully received
    Serial.println(F("[[[NEW PACKET]]]"));

    Serial.print(F("Time:\t\t\t"));
    Serial.println(millis());

    Serial.print(F("Data:\t\t\t"));
    Serial.println(str);

    Serial.print(F("RSSI:\t\t\t"));
    Serial.print(radio.getRSSI());
    Serial.println(F(" dBm"));

    Serial.print(F("SNR:\t\t\t"));
    Serial.print(radio.getSNR());
    Serial.println(F(" dB"));

    Serial.print(F("FreqError:\t\t"));
    Serial.print(radio.getFrequencyError());
    Serial.println(F(" Hz"));

    Serial.print(F("Freq:\t\t\t"));
    Serial.print(frequency[freqCounter % freqSize]);
    Serial.println(F(" Hz"));

    Serial.print(F("SF:\t\t\t"));
    Serial.println(sf[sfCounter % sfSize]);

    Serial.println();
  } else if (state == RADIOLIB_ERR_RX_TIMEOUT) {
    // timeout occurred while waiting for a packet
    // Serial.println(F("timeout!"));     IGNORE
  } else if (state == RADIOLIB_ERR_CRC_MISMATCH) {
    // packet was received, but is malformed
    Serial.println(F("CRC error!"));
  } else {
    // some other error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);
  }

  bool changeFreq = false;
  sfCounter++;

  if (sfCounter % sfSize == 0) {
    freqCounter++;
    changeFreq = true;
  }

  float newFreq = frequency[freqCounter % freqSize];
  int newSf = sf[sfCounter % sfSize];

  if (changeFreq && freqSize != 1) {
    if (radio.setFrequency(newFreq) == RADIOLIB_ERR_INVALID_FREQUENCY) {
      Serial.println(F("Selected frequency is invalid for this module!"));
      return;
    }

    // Serial.print(F("Set Freq to "));
    // Serial.println(newFreq);
  }

  if (radio.setSpreadingFactor(newSf) == RADIOLIB_ERR_INVALID_SPREADING_FACTOR) {
    Serial.println(F("Selected spreading factor is invalid for this module!"));
    return;
  }
}

```

**Comment by jgromes** [comment, marked as answer]
> Using int state = radio.receive(str); I get RADIOLIB_ERR_NONE, but no data is displayed

From your code, you seem to be attempting to print the data as String - raw LoRaWAN packets are binary and may not contain any printable characters, especiallyy since a part of the packet will be encrpyted most of the time. So you are actually receiving a packet (because the return is `RADIOLIB_ERR_NONE`, it would have been `RADIOLIB_ERR_RX_TIMEOUT` otherwise), you're just not printing it correctly.

However, there's a more fundamental issue you will run into - LoRaWAN channels may change dynamically: the server may update the node frequency, spreading factor or bandwidth based on internal logic that your sniffer cannot predict. With SX126x, you cannot cover all the possible frequencies the server could choose, that's why gateways use a lot more complex transceivers than the SX126x. What you could do is use channel activity detection instead of receiveing, and only going to receive mode if CAD actually detects some LoRa packet, but since for each freqeuency, you are dealing with multiple different SF/BW configurations, I doubt it will be fast enough to actually catch anything. As an example, EU868 has 3 default frequency channels and 8 possible data rates, so that's 24 configurations at the very least (there may be more frequency channels defined while joining the network). And it should go without saying, you will not be able to get any payload data provided the LoRaWAN network is set up correctly.

**Comment by joterr** [comment]
How stupid. I'm really new to the Arduino world -  I thought the print command converts the data correctly without further specification.
Now I'm receiving the packet data as a byte array and loop the values to generate a HEX string. Thank you very much!

I am aware that there are several channels and that the SX126x can only listen to one of them. I tried iterating over the common OTAA and data channels to be able to sniff some packets. The limitations are acceptable for my purpose.

This structure is only for a practical part of my master's thesis, in which I want to understand the weak points of LoRaWAN. 


```
  byte byteArr[packetLength];
  int state = radio.receive(byteArr, packetLength);

 Serial.print(F("Data (HEX):\t\t"));

      for (int i = 0; i < packetLength; i++) {
        if (byteArr[i] < 0x10) {
          hexstring += '0';
        }

        hexstring += String(byteArr[i], HEX);
      }

      Serial.print(hexstring);
```
